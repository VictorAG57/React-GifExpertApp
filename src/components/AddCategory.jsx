import React, { useState } from 'react';
import PropsTypes from "prop-types";

/** Función del componente AddCategory
 * 
 * Esté me permite agregar una categoria a mi página, el usuario ingresa
 * en una caja de texto la categoría que desea agragar, y al precionar enter (submit)
 * actualiza el estado y lo agrega a una lista con las demás categorias.
 * 
 * Esté componente complementa al componente pruncipal GifExpertApp.
 */

export const AddCategory = ({ setCategory }) => {

    const [inputValue, setInputValue] = useState("")

    //Obtenemos el valor de nuestra caja de texto para actualizar el estado.
    //Lo escrito por el usuario
    const handleInputChange = (e)=> {
        setInputValue(e.target.value)
        console.log("handleInputChange llamado")
    };

    //Cambiamos el estado con "handleSubmit", agregamos un elemento más al etsado
    const handleSubmit = (e)=> {
        //Evitamos que el usuario al dar enter refresque la página
        //ya que es un formulario
        e.preventDefault()

        //Meidante el if, evitamos que se pueda agregar a la lista un elemento
        //sin coontenido
        if(inputValue.trim().length > 2){
            //Primero colocamos inputValue, para cuando se renderize, lo
            //primero que aparezac sea la ultima busqueda y no aparezca hasta el final
            setCategory( cats => [ inputValue, ...cats])
            setInputValue("")
            console.log("setCategory llamado")
        }
    };

    return (
        //Usamos el onSubmit pra implementar la función "handleSubmit"
        <form onSubmit={ handleSubmit }>
            <input type="text" 
                value={ inputValue } 
                onChange={ handleInputChange }>
            </input>
        </form>
    );
}

//Obligamos a que "setCategory" sea obligatoria
AddCategory.propTypes = {
    setCategory: PropsTypes.func.isRequired
};

