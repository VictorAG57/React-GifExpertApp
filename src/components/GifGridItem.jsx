import React from 'react';
import "../index.css";
import PropsTypes from "prop-types";
import { prototype } from 'webpack-sources/lib/CachedSource';


/**
 * EstÃ© componente es el encargado de la estructura de cada
 * carta de gif que aparece.
 */

export const GifGridItem = ({ id, title, url}) => {
    
    // console.log(img)

    return (
        <div className="card animate__animated animate__bounce animate__delay-2s animate_fadeIn">
            <img src={ url } alt={ title }></img>
            <p>{ title }</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropsTypes.string.isRequired,
    url: PropsTypes.string.isRequired
}