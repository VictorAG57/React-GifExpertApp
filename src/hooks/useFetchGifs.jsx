import { useState, useEffect } from "react";
import { GetGifs } from "../helpers/GetGifs";

/** CUSTOM HOOK
 * 
 * Esté es un custom hook que nos permite usar la data traida de mi
 * helper GetGifs, para realiar una acción secundaria mediante useEffect.
 * Aquí actualizamos el estado para solamente enviar la información
 * a mi componente de grilla de cursos, y ahi lo renderizen.
 * 
 */

export const useFetchGifs = ( category )=> {
    //Usaos useEffect como acciones secundarias, y aqui la usamos con el fin
    //de que una vez cargue la página realiza la petición http para traer el contenido.

    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    useEffect( ()=> {

        setTimeout( ()=> {
            GetGifs( category )
                .then(img => {
                    setstate({
                        data: img,
                        loading: false
                    })
                })
        }, 2000)
    }, [ category ])

    return state;

}