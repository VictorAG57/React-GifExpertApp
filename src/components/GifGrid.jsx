import React, { Fragment} from 'react';
import PropTypes from "prop-types";

import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from "./GifGridItem";

//Se encarga de la estructura de cada Item de gif, para acomodarla
//en una estructutura, los organiza todos, para enviar la infotmación
//a GifExpertApp y la renderize.

//Resive omo argumento la categoría enviada por el usuario para que useFetchGifs
//se encargue de estructurar la información, que as u vez se conecta con nuestro helper
//para realizar la petición.

export const GifGrid = ({ category }) => {

    //useFetchGifs me traera la información de mis gifs
    const { data:images, loading } = useFetchGifs( category )

    return (
        <Fragment>
            <h3 className="animate__animated animate__bounce animate__delay-2s">{ category }</h3>

            { loading && <p className="animate__fadeInRight" >Loading</p> }

            <div className="card-grid">
                { //También puedes extraer las propiedades en img --> {id, title}
                //Manda como parametros la url:img, y el titulo:id a nuestro <GifGridItem>
                images.map( img => ( 
                    <GifGridItem
                        key={img.id} 
                        {...img}>
                    </GifGridItem> 
                    ))
                }
            </div> 
           
        </Fragment>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

// const [images, setImages] = useState([])

    //Usaos useEffect como acciones secundarias, y aqui la usamos con el fin
    //de que una vez cargue la página realiza la petición http para traer el contenido.
    
    // useEffect( ()=> {
    //     GetGifs( category )
    //         .then(imgs => setImages( imgs ))
    // }, [ category ]) 

    //Colocamos [category] como argumento, por si es 
    //necesario que useEffect se vuelva a renderizar, aunque no vaya a pazar pero
    //así eliminamos un warning que nos aparecera en cosola.

 
    /*
        <ol>
            { //También puedes extraer las propiedades en img --> {id, title}
                images.map( img => (<li key={ img.id }> { img.title } </li>) )
            }
        </ol>
    */