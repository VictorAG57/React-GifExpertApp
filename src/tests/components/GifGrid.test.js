import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
import '@testing-library/jest-dom';

//Me permite simular la llamada de la función, y controlamos la 
//información que se respondería.
jest.mock("../../hooks/useFetchGifs")

describe('Debe de renderizar el componente <GifGrid></GifGrid>', () => {

    const category = "Rick and Morty"

    useFetchGifs.mockReturnValue({
        data: [], 
        loading: true
    })

    test('Debe de mostrar correctamente el componente y snapShot ', () => {

        //Al testear GifGrid, también es necesario evaluar que dentro de el, llama nuestro
        //useFetchGifs, por ello simulamos esa función jest.mock("../../hooks/useFetchGifs"),
        //y le damos un valor inicial para que pueda funcionar.

        /**  mockReturn() 
         * lo que hace es que cuando se llame este fetch dentro de mi componente,
         * lo que va a hacer es regresar este valor y dentro de estos paréntesis van a colocar unas llaves que
         * va a ser el objeto que va a tomar mi componente acá.
         * Es decir, va a ser el resultado de esto ahora.
        */

        useFetchGifs.mockReturnValue({
            // Valores por defecto de mi useFerchGifs, propios de nuestro customHook
            data: [],     // La data que obtiene nuestro useFetchGifs
            loading: true // El valor de carga
        })

        const wrapper = shallow(<GifGrid category={ category }></GifGrid>)

        expect( wrapper ).toMatchSnapshot() 
    });

    test('Debe de mostrar items cuando se carga useFetchGifs (<GifGrid></GifGrid>)', () => {

        // Simulamos el valor de respuesta de nuestra data obtenida de useFetchGifs
        const gifs = [{
            id: "ABC",
            url: "https://lago.jpg",
            title: "Hola mundo"
        }]
        
        // Aquí simulamos que nuestro customHook a traido la siguiente información
        useFetchGifs.mockReturnValue({
            data: gifs,   // La data que obtiene nuestro useFetchGifs
            loading: true // El valor de carga
        });
        
        const wrapper = shallow(<GifGrid category={ category }></GifGrid>)

        // Una evaliación sencilla 
        expect( wrapper ).toMatchSnapshot()

        // Evaluamos si existe código con la etiqueta 'p'
        // código que no deberia de existir.
        expect( wrapper.find('spam').exists()).toBe(false);

        // Nuestro GigGrid debe contener dentro la llamada de un componete llamado 'GifGridItem'
        // y aquí evaluamos cuantos elementos tiene dentro (1 arreglo), 
        // que deben ser los mismos que nuestra data:gifs
        // que es la que le enviamos a dicho conponente
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length )
        console.log(gifs.length)
    });
});
