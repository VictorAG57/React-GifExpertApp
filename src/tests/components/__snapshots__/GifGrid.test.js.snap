// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debe de renderizar el componente <GifGrid></GifGrid> Debe de mostrar correctamente el componente y snapShot  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounce animate__delay-2s"
  >
    Rick and Morty
  </h3>
  <p
    className="animate__fadeInRight"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Debe de renderizar el componente <GifGrid></GifGrid> Debe de mostrar items cuando se carga useFetchGifs (<GifGrid></GifGrid>) 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounce animate__delay-2s"
  >
    Rick and Morty
  </h3>
  <p
    className="animate__fadeInRight"
  >
    Loading
  </p>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Hola mundo"
      url="https://lago.jpg"
    />
  </div>
</Fragment>
`;
